src/content_scripts/autocorrect.js

```diff
@@ -0,0 +1,400 @@
+"use strict";
+
+const fractions = Object.freeze({
+    "¼": 1.0 / 4.0,
+    "½": 1.0 / 2.0,
+    "¾": 3.0 / 4.0,
+    "⅐": 1.0 / 7.0,
+    "⅑": 1.0 / 9.0,
+    "⅒": 1.0 / 10.0,
+    "⅓": 1.0 / 3.0,
+    "⅔": 2.0 / 3.0,
+    "⅕": 1.0 / 5.0,
+    "⅖": 2.0 / 5.0,
+    "⅗": 3.0 / 5.0,
+    "⅘": 4.0 / 5.0,
+    "⅙": 1.0 / 6.0,
+    "⅚": 5.0 / 6.0,
+    "⅛": 1.0 / 8.0,
+    "⅜": 3.0 / 8.0,
+    "⅝": 5.0 / 8.0,
+    "⅞": 7.0 / 8.0
+});
+
+const constants = Object.freeze({
+    "π": Math.PI,
+    "e": Math.E
+});
+
+// communication type
+// directly include magic constant as a workaround as we cannot import modules in content scripts due to https://bugzilla.mozilla.org/show_bug.cgi?id=1451545
+const AUTOCORRECT_CONTENT = "autocorrectContent";
+
+let insertedText; // Last insert text
+let deletedText; // Last deleted text
+let lastTarget; // Last target
+let lastCaretPosition; // Last caret position
+
+let quotes = true;
+let fracts = true;
+
+let autocorrections = {};
+
+let longest = 0;
+
+// Regular expressions
+let symbolpatterns = null;
+// Do not autocorrect for these patterns
+let apatterns = null;
+
+// Thunderbird
+const THUNDERBIRD = typeof messenger !== "undefined";
+
+// Chrome
+const CHROME = Object.getPrototypeOf(browser) !== Object.prototype;
```

Comment:
Ah we cannot use modules here yet, can we?
Because otherwise I would have said, we should put that part in it's own module/helper.

Also we'd better use [the specific WebExtenion APIs for that](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/runtime).
Or is this feature-checking? Because I do not see at all how this detects Thunderbird or Chrome/ium?

So if you do browser sniffing, let's use the proper APIs we have for this.
If you do feature sniffing – which I agree is the better thing in general, then do name the constants you define here like that.

In any case, some comments explaining *why*/*how* this detects things may be useful. Especially, if you have, links to the bug/feature issues/reports/tickets, where this missing feature has been requested.

Reply:
No, content scripts cannot use modules or [those Web Extension Runtime APIs](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Content_scripts#WebExtension_APIs).

For Thunderbird (quoted from [here](https://thunderbird-webextensions.readthedocs.io/en/latest/)):
> In Thunderbird, all WebExtension API can be accessed through the browser.* namespace, as with Firefox, but also through the messenger.* namespace, which is a better fit for Thunderbird.

I believe checking for this namespace is the best way to reliably detect Thunderbird. This is why we needed [bug 1641573](https://bugzilla.mozilla.org/show_bug.cgi?id=1641573) to be fixed.

For Chrome, yes that is doing feature detection. Specifically it is detecting the webextension-polyfill, which is currently needed for all Chromium based browsers. It was copied from [this line](https://github.com/mozilla/webextension-polyfill/blob/master/src/browser-polyfill.js#L9) of the webextension-polyfill itself.

I will add some comments to the code with these links.

Reply:
Yeah, note if we can avoid those checks at all (and fallback to something when we e.g. check for the result), we should of course.